src/main/java/org/aksw/iguana/cc/worker/impl/SPARQLProtocolWorker.java

```diff
@@ -0,0 +1,412 @@
+                      Boolean parseResults) {
+            this.number = number == null ? 1 : number;
+            this.queries = queries;
+            this.completionTarget = completionTarget;
+            this.connection = connection;
+            this.timeout = timeout;
+            this.acceptHeader = acceptHeader;
+            this.requestType = requestType == null ? RequestFactory.RequestType.GET_QUERY : requestType;
+            this.parseResults = parseResults == null || parseResults;
+        }
+    }
+
+    record HttpExecutionResult(
+            int queryID,
+            Optional<HttpResponse<InputStream>> response,
+            Instant requestStart,
+            Duration duration,
+            Optional<BigByteArrayOutputStream> outputStream,
+            OptionalLong actualContentLength,
+            OptionalLong hash,
+            Optional<Exception> exception
+    ) {
+        public boolean completed() {
+            return response.isPresent();
+        }
+
+        public boolean successful() {
+            if (completed() && exception.isEmpty())
+                return (response.get().statusCode() / 100) == 2;
+            return false;
+        }
+    }
+
+
+    private HttpClient httpClient;
+    private final ThreadPoolExecutor executor;
+
+    private final XXHashFactory hasherFactory = XXHashFactory.fastestJavaInstance();
+    private final RequestFactory requestFactory;
+
+    private final ResponseBodyProcessor responseBodyProcessor;
+
+    // declared here, so it can be reused across multiple method calls
+    private BigByteArrayOutputStream responseBodybbaos = new BigByteArrayOutputStream();
+
+    // used to read the http response body
+    private final byte[] buffer = new byte[4096];
+
+    private final static Logger LOGGER = LoggerFactory.getLogger(SPARQLProtocolWorker.class);
+
+    @Override
+    public Config config() {
+        return (SPARQLProtocolWorker.Config) config;
+    }
+
+
+    public SPARQLProtocolWorker(long workerId, ResponseBodyProcessor responseBodyProcessor, Config config) {
+        super(workerId, responseBodyProcessor, config);
+        this.responseBodyProcessor = responseBodyProcessor;
+        this.executor = (ThreadPoolExecutor) Executors.newFixedThreadPool(1);
+        this.requestFactory = new RequestFactory(config().requestType());
+        this.httpClient = buildHttpClient();
+    }
+
+
+    public CompletableFuture<Result> start() {
+        return CompletableFuture.supplyAsync(() -> {
+            List<ExecutionStats> executionStats = new ArrayList<>();
+            if (config().completionTarget() instanceof QueryMixes queryMixes) {
+                for (int i = 0; i < queryMixes.number(); i++) {
+                    for (int j = 0; j < config().queries().getQueryCount(); j++) {
+                        ExecutionStats execution = executeQuery(config().timeout(), false);
+                        if (execution != null) {
```

Comment:
execution should never be null here, right? better use an assertion then. Or even throw if null